base/generator.jl

```diff
@@ -20,4 +22,37 @@ function next(g::Generator, s)
     g.f(v), s2
 end
 
-collect(g::Generator) = map(g.f, g.iter)
+## iterator traits
+
+abstract IteratorSize
+immutable SizeUnknown <: IteratorSize end
+immutable HasLength <: IteratorSize end
+immutable HasShape <: IteratorSize end
+immutable IsInfinite <: IteratorSize end
+
+iteratorsize(x) = iteratorsize(typeof(x))
+iteratorsize(::Type) = HasLength()  # HasLength is the default
+
+and_iteratorsize{T}(isz::T, ::T) = isz
+and_iteratorsize(::HasLength, ::HasShape) = HasLength()
+and_iteratorsize(::HasShape, ::HasLength) = HasLength()
+and_iteratorsize(a, b) = SizeUnknown()
+
+abstract IteratorEltype
+immutable EltypeUnknown <: IteratorEltype end
+immutable HasEltype <: IteratorEltype end
+
+iteratoreltype(x) = iteratoreltype(typeof(x))
```

Comment:
Merit in the `iterator` prefix, too. `iteratorhaseltype`? Getting kinda long without underscores, though.